tests/src/python/test_provider_ogr.py

```diff
@@ -2247,6 +2247,72 @@ def test_provider_sidecar_files_for_uri(self):
                          ['/home/me/special.gfs', '/home/me/special.xsd'])
         self.assertEqual(metadata.sidecarFilesForUri('/home/me/special.csv'), ['/home/me/special.csvt'])
 
+    def testGeoJsonFieldOrder(self):
+        """Test issue GH #45139"""
+
+        d = QTemporaryDir()
+        json_path = os.path.join(d.path(), 'test.geojson')
+        with open(json_path, 'w+') as f:
+            f.write("""
+            {
+                "type": "FeatureCollection",
+                "features": [
+                    {
+                        "type": "Feature",
+                        "geometry": {
+                            "type": "Point",
+                            "coordinates": [11.1215698,46.0677293]
+                        },
+                        "properties": {
+                            "A": "A",
+                        }
+                    },
+                    {
+                        "type": "Feature",
+                        "geometry": {
+                            "type": "Point",
+                            "coordinates": [11.1214686,46.0677385]
+                        },
+                        "properties": {
+                            "A": "A",
+                            "B": "B",
+                        }
+                    }
+                ]
+            }
+            """)
+
+        vl = QgsVectorLayer(json_path, 'json')
+        self.assertTrue(vl.isValid())
+        self.assertEqual(vl.featureCount(), 2)
+        self.assertEqual(vl.fields().names(), ['A', 'B'])
+
+        # Append a field
+        self.assertTrue(vl.startEditing())
+        self.assertTrue(vl.addAttribute(QgsField('C', QVariant.String)))
+
+        for f in vl.getFeatures():
+            vl.changeAttributeValue(f.id(), 2, 'C')
+
+        self.assertEqual(vl.fields().names(), ['A', 'B', 'C'])
+
+        features = [f for f in vl.getFeatures()]
+
+        self.assertEqual(features[0].attribute('B'), NULL)
+        self.assertEqual(features[0].attribute('C'), 'C')
+        self.assertEqual(features[1].attribute('B'), 'B')
+        self.assertEqual(features[1].attribute('C'), 'C')
+
+        self.assertTrue(vl.commitChanges())
+        self.assertEqual(vl.fields().names(), ['A', 'C', 'B'])
```

Comment:
this will no longer be true in GDAL 3.4 since https://github.com/OSGeo/gdal/pull/4552.  So this whole test should be skipped for GDAL >= 3.4, or modified to handle the A, B, C order as well.